File: Sorting/Radix Sort/RadixSort.py
# Function to get the largest element from the array
def get_max(arr):
    return max(arr)

# Counting sort to sort elements based on the digit represented by `place`
def counting_sort(arr, place):
    base = 10  # Base of the number system
    n = len(arr)
    output = [0] * n  # Output array
    count = [0] * base  # Count array

    # Count occurrences of each digit at the given place
    for num in arr:
        digit = (num // place) % base
        count[digit] += 1

    # Calculate cumulative count to determine positions
    for i in range(1, base):
        count[i] += count[i - 1]

    # Build the output array
    for i in range(n - 1, -1, -1):
        digit = (arr[i] // place) % base
        output[count[digit] - 1] = arr[i]
        count[digit] -= 1

    # Copy the sorted elements back to the original array
    for i in range(n):
        arr[i] = output[i]

# Main function to implement Radix Sort
def radix_sort(arr):
    # Get the maximum number to determine the number of digits
    max_num = get_max(arr)

    # Perform counting sort for each digit, starting from least significant
    place = 1
    while max_num // place > 0:
        counting_sort(arr, place)
        place *= 10

# Function to display the array
def display(arr):
    print("\t".join(map(str, arr)))

# Demonstration of the radix sort
arr = [112, 400, 543, 441, 678, 675, 9, 777]
print("Before sorting:")
display(arr)

radix_sort(arr)

print("After sorting:")
display(arr)

File: Sorting/Radix Sort/RadixSort.cpp
#include <iostream>
using namespace std;

// Function to get the largest element from an array
int getMax(int array[], int n) {
    int max = array[0];
    for (int i = 1; i < n; i++) {
        if (array[i] > max)
            max = array[i];
    }
    return max;
}

// Counting sort function to sort elements based on the digit represented by `place`
void countSort(int array[], int size, int place) {
    const int base = 10; // Base of the number system (decimal)
    int output[size];    // Output array to store sorted elements
    int count[base] = {0}; // Initialize count array to zero

    // Count occurrences of each digit at the given place value
    for (int i = 0; i < size; i++) {
        int digit = (array[i] / place) % base;
        count[digit]++;
    }

    // Calculate cumulative count to determine the position of each digit
    for (int i = 1; i < base; i++) {
        count[i] += count[i - 1];
    }

    // Build the output array by placing elements in sorted order
    for (int i = size - 1; i >= 0; i--) {
        int digit = (array[i] / place) % base;
        output[count[digit] - 1] = array[i];
        count[digit]--;
    }

    // Copy the sorted elements back to the original array
    for (int i = 0; i < size; i++) {
        array[i] = output[i];
    }
}

// Main function to implement Radix Sort
void radixSort(int array[], int size) {
    // Get the maximum number to determine the number of digits
    int max = getMax(array, size);

    // Perform counting sort for each digit, moving from least significant to most significant
    for (int place = 1; max / place > 0; place *= 10) {
        countSort(array, size, place);
    }
}

// Function to display the elements of an array
void display(int array[], int size) {
    for (int i = 0; i < size; i++) {
        cout << array[i] << "\t";
    }
    cout << endl;
}

// Driver code
int main() {
    int array[] = {112, 400, 543, 441, 678, 675, 9, 777};
    int n = sizeof(array) / sizeof(array[0]);

    cout << "Before sorting:" << endl;
    display(array, n);

    radixSort(array, n);

    cout << "After sorting:" << endl;
    display(array, n);

    return 0;
}
